src/inc/til/small_vector.h

```diff
@@ -0,0 +1,747 @@
+        {
+            return std::equal(begin(), end(), other.begin(), other.end());
+        }
+
+        reference front() noexcept
+        {
+            return operator[](0);
+        }
+
+        const_reference front() const noexcept
+        {
+            return operator[](0);
+        }
+
+        reference back() noexcept
+        {
+            return operator[](_size - 1);
+        }
+
+        const_reference back() const noexcept
+        {
+            return operator[](_size - 1);
+        }
+
+        reference operator[](size_type off) noexcept
+        {
+            assert(off < _size);
+            return _data[off];
+        }
+
+        const_reference operator[](size_type off) const noexcept
+        {
+            assert(off < _size);
+            return _data[off];
+        }
+
+        reference at(size_type off)
+        {
+            if (off >= _size)
+            {
+                _throw_invalid_subscript();
+            }
+            return _data[off];
+        }
+
+        const_reference at(size_type off) const
+        {
+            if (off >= _size)
+            {
+                _throw_invalid_subscript();
+            }
+            return _data[off];
+        }
+
+        void clear() noexcept
+        {
+            std::destroy_n(begin(), _size);
+
+            if (_capacity != N)
+            {
+                delete[] _data;
+            }
+
+            _data = &_buffer[0];
+            _capacity = N;
+            _size = 0;
+        }
+
+        void resize(size_type newSize)
+        {
+            if (newSize < _size)
+            {
+                std::destroy_n(_data + newSize, _size - newSize);
+            }
+            else if (newSize > _size)
+            {
+                if (newSize > _capacity)
+                {
+                    _grow(newSize - _capacity);
+                }
+
+                std::uninitialized_value_construct_n(_data + _size, newSize - _size);
+            }
+
+            _size = newSize;
+        }
+
+        void resize(size_type newSize, const_reference value)
+        {
+            if (newSize < _size)
+            {
+                std::destroy_n(_data + newSize, _size - newSize);
+            }
+            else if (newSize > _size)
+            {
+                if (newSize > _capacity)
+                {
+                    _grow(newSize - _capacity);
+                }
+
+                std::uninitialized_fill_n(_data + _size, newSize - _size, value);
+            }
+
+            _size = newSize;
+        }
+
+        void shrink_to_fit()
+        {
+            if (_capacity == N || _size == _capacity)
+            {
+                return;
+            }
+
+            auto data = &_buffer[0];
+            if (_size > N)
+            {
+                data = new T[_size];
+            }
+
+            std::uninitialized_move_n(begin(), _size, data);
+            std::free(_data);

```

Comment:
i see that sometimes you use `std::free` and sometimes you use `delete[]`.. which I think is incorrect, right?

Reply:
I just realized that the use of `new`/`delete` is entirely incorrect. They initialize the objects, which we explicitly don't want to for a vector.